js/gacha.js

```diff
@@ -1,10 +1,99 @@
 import { renderCards } from "./cards.js";
+import { cards_by_rarity } from "./main.js";
 
 export const GACHA_BUTTON = document.getElementById("gacha-button");
-export const CARDS_PER_PULL = 10;
 
-//Pulls n cards from the passed "cards" array using the Fisher-Yates Shuffle.
-function pullCards(cards, n) {
+//Card slots on each pull, representing the percent chance of getting a card of that rarity in each slot.
+//The sum of all rarities on a slot should be 100 or higher for proper function.
+const slots = [
+  {
+    Common: 100,
+  },
+  {
+    Common: 100,
+  },
+  {
+    Common: 80,
+    Uncommon: 20,
+  },
+  {
+    Common: 80,
+    Uncommon: 20,
+  },
+  {
+    Uncommon: 100,
+  },
+  {
+    Uncommon: 100,
+  },
+  {
+    Uncommon: 26,
+    Rare: 26,
+    HoloRare: 21,
+    UltraRare: 17,
+    SecretRare: 10,
+  },
+  {
+    Uncommon: 26,
+    Rare: 26,
+    HoloRare: 21,
+    UltraRare: 17,
+    SecretRare: 10,
+  },
+  {
+    Rare: 40,
+    HoloRare: 20,
+    UltraRare: 20,
+    SecretRare: 20,
+  },
+  {
+    Rare: 40,
+    HoloRare: 20,
+    UltraRare: 20,
+    SecretRare: 20,
+  },
+];
+
+//Slots that replace one or more normal slots on special conditions.
+//This can be used to help collect all cards.
+const specialSlots = {
+  69: {
+    HoloRare: 100 / 3,
+    UltraRare: 100 / 3,
+    SecretRare: 100 / 3,
+  },
+};
+
+//Pulls a set of cards guided by an array of "slots" passed to the function.
+function pullCards(slots) {
+  let cards = [];
+  for (let slot of slots) {
+    let dice = Math.floor(Math.random() * 100) + 1;
+    if (dice in specialSlots) {
+      slot = specialSlots[dice];
+    }
```

Comment:
does this actually occur in the real TCG rate system? :0

Reply:
Yes! Take a look at the official page for Booster Pack Slots: [https://tcg.pokemon.com/en-us/tcgl/droprate/](https://tcg.pokemon.com/en-us/tcgl/droprate/)
![Slots](https://user-images.githubusercontent.com/33675019/209709797-fff6da4c-37db-40fd-b7ad-5ee04ec69da0.png)
These are the slots for the "Sword & Shield-Lost Origin" expansion. The ones on the top row (Available Rarity) are the normal slots, and the ones on the bottom row (Available Upgrade Rarity) are special slots that have a small chance of replacing the standard drop in that slot. I imagine it brings more excitement to the pulls, and it also helps those who open them in getting rare cards.